src/praisonai-agents/praisonaiagents/runtime/doctor_registry.py

```
@@ -0,0 +1,147 @@
+        self._rules: List[DoctorContractProtocol] = []
+        self._plugins_loaded = False
+    
+    def register_rule(self, rule: DoctorContractProtocol) -> None:
+        """Register a migration rule."""
+        if not isinstance(rule, DoctorContractProtocol):
+            raise TypeError("Rule must implement DoctorContractProtocol")
+        
+        # Check for duplicate rule IDs
+        for existing_rule in self._rules:
+            if existing_rule.rule_id == rule.rule_id:
+                warnings.warn(f"Duplicate rule ID '{rule.rule_id}' - replacing existing rule")
```

Ruff (0.15.18)

[warning] No explicit `stacklevel` keyword argument found

Set `stacklevel=2`

(B028)